File: node/director/telemetry-workflow/handler.js
'use strict'

const { performance } = require('node:perf_hooks')

const gatewayURL = process.env.gateway_url ||
  'http://gateway.openfaas:8080'
const stageTimeout = configuredTimeout(
  process.env.stage_timeout || '5',
  'stage_timeout'
)

module.exports = async (event, context) => {
  const started = performance.now()
  const input = requestBody(event.body)

  let validated
  try {
    validated = await invoke('validate-reading', input)
  } catch (error) {
    return fail(
      context,
      502,
      `failed to invoke validate-reading: ${error.message}`
    )
  }

  if (validated.status !== 200) {
    return fail(
      context,
      validated.status,
      `validate-reading failed: ${await validated.text()}`
    )
  }

  let reading
  try {
    reading = await validated.json()
  } catch (error) {
    return fail(
      context,
      502,
      `unexpected response from validate-reading: ${error.message}`
    )
  }

  const body = JSON.stringify(reading)
  const names = ['temperature-check', 'battery-check']
  let responses
  try {
    responses = await Promise.all(
      names.map(async (name) => [name, await invoke(name, body)])
    )
  } catch (error) {
    return fail(context, 502, `failed to invoke check: ${error.message}`)
  }

  const completed = Object.fromEntries(responses)
  for (const name of names) {
    const response = completed[name]
    if (response.status !== 200) {
      return fail(
        context,
        response.status,
        `${name} failed: ${await response.text()}`
      )
    }
  }

  let temperature
  let battery
  try {
    temperature = await completed['temperature-check'].json()
    battery = await completed['battery-check'].json()
  } catch (error) {
    return fail(
      context,
      502,
      `unexpected response from check function: ${error.message}`
    )
  }

  const status = temperature.alert || battery.alert ? 'alert' : 'ok'
  return context
    .status(200)
    .headers({ 'Content-Type': 'application/json' })
    .succeed({
      device_id: reading.device_id,
      status,
      temperature,
      battery,
      duration_ms: Math.round(performance.now() - started)
    })
}

function invoke (name, body) {
  const gateway = gatewayURL.replace(/\/$/, '')
  return fetch(`${gateway}/function/${name}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body,
    signal: AbortSignal.timeout(stageTimeout)
  })
}

function requestBody (body) {
  if (Buffer.isBuffer(body)) {
    return body
  }
  return typeof body === 'string' ? body : JSON.stringify(body)
}

function configuredTimeout (value, name) {
  const seconds = Number(value)
  if (!Number.isFinite(seconds) || seconds <= 0) {
    throw new Error(`${name} must be greater than zero`)
  }
  return seconds * 1000
}

function fail (context, status, message) {
  return context
    .status(status)
    .headers({ 'Content-Type': 'text/plain' })
    .succeed(message.trim())
}

File: python/director/telemetry-workflow/handler.py
import concurrent.futures
import os
import time

import requests


GATEWAY_URL = os.getenv(
    "gateway_url", "http://gateway.openfaas:8080"
).rstrip("/")
STAGE_TIMEOUT = float(os.getenv("stage_timeout", "5"))
if STAGE_TIMEOUT <= 0:
    raise ValueError("stage_timeout must be greater than zero")


def handle(event, context):
    started = time.monotonic()
    body = event.body

    try:
        validated = invoke("validate-reading", body)
    except requests.RequestException as err:
        return error(502, f"failed to invoke validate-reading: {err}")

    if validated.status_code != 200:
        return error(
            validated.status_code,
            f"validate-reading failed: {validated.text}",
        )

    try:
        reading = validated.json()
    except ValueError as err:
        return error(502, f"unexpected response from validate-reading: {err}")

    functions = ("temperature-check", "battery-check")
    completed = {}
    with concurrent.futures.ThreadPoolExecutor(max_workers=2) as executor:
        futures = {
            name: executor.submit(invoke, name, validated.content)
            for name in functions
        }
        for name, future in futures.items():
            try:
                completed[name] = future.result()
            except requests.RequestException as err:
                return error(502, f"failed to invoke {name}: {err}")

    for name in functions:
        response = completed[name]
        if response.status_code != 200:
            return error(
                response.status_code,
                f"{name} failed: {response.text}",
            )

    try:
        temperature = completed["temperature-check"].json()
        battery = completed["battery-check"].json()
    except ValueError as err:
        return error(502, f"unexpected response from check function: {err}")

    has_alert = temperature["alert"] or battery["alert"]
    workflow_status = "alert" if has_alert else "ok"
    return {
        "statusCode": 200,
        "body": {
            "device_id": reading["device_id"],
            "status": workflow_status,
            "temperature": temperature,
            "battery": battery,
            "duration_ms": int((time.monotonic() - started) * 1000),
        },
    }


def invoke(function, body):
    return requests.post(
        f"{GATEWAY_URL}/function/{function}",
        data=body,
        headers={"Content-Type": "application/json"},
        timeout=STAGE_TIMEOUT,
    )


def error(status_code, message):
    return {"statusCode": status_code, "body": message.strip()}
